apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+    */
+    function _canChallenge(uint256 _actionId, Action storage _action) internal view returns (bool) {
+        return _canProceed(_actionId, _action) && _action.disputable.canChallenge(_action.disputableActionId);
+    }
+
+    /**
+    * @dev Tell whether an action can be closed
+    * @param _actionId Identification number of the action to be queried
+    * @param _action Action instance to be queried
+    * @return True if the action can be closed, false otherwise
+    */
+    function _canClose(uint256 _actionId, Action storage _action) internal view returns (bool) {
+        return _canProceed(_actionId, _action) && _action.disputable.canClose(_action.disputableActionId);
+    }
+
+    /**
+    * @dev Tell whether an action can proceed.
+    * @dev An action can proceed if it is not:
+    * @dev  - Closed
+    * @dev  - Currently challenged
+    * @dev  - Ruled as voided
+    * @dev  - Ruled in favour of the submitter
+    * @param _actionId Identification number of the action to be queried
+    * @param _action Action instance to be queried
+    * @return True if the action can proceed, false otherwise
+    */
+    function _canProceed(uint256 _actionId, Action storage _action) internal view returns (bool) {
+        // If the action was already closed, return false
+        if (_action.closed) {
+            return false;
+        }
+
+        uint256 challengeId = _action.currentChallengeId;
+        Challenge storage challenge = challenges[challengeId];
+
+        // If the action was not challenged, return true
+        if (!_existChallenge(_actionId, challengeId, challenge)) {
+            return true;
+        }
+
+        // If the action was challenged but ruled in favour of the submitter
+        // (dispute rejected by arbitrator) or voided, return true
+        ChallengeState state = challenge.state;
+        return state == ChallengeState.Rejected || state == ChallengeState.Voided;
+    }
+
+    /**
+    * @dev Tell whether an action can be settled
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action can be settled, false otherwise
+    */
+    function _canSettle(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        return _isWaitingChallengeAnswer(_actionId, _challenge);
+    }
+
+    /**
+    * @dev Tell whether an action can be disputed
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenge Current challenge instance for the action
+    * @return True if the action can be disputed, false otherwise
+    */
+    function _canDispute(uint256 _actionId, Challenge storage _challenge) internal view returns (bool) {
+        if (!_isWaitingChallengeAnswer(_actionId, _challenge)) {
+            return false;
+        }
+
+        return _challenge.endDate > getTimestamp64();
```

Comment:
In cases where we're comparing timestamps, we should cast up to uint256 as it's a bit cheaper.